packages/fx-core/src/component/driver/arm/util/handleError.ts

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { ResourceManagementClient } from "@azure/arm-resources";
+import { err, FxError, ok, Result, SolutionContext } from "@microsoft/teamsfx-api";
+import { SolutionError } from "../../../constants";
+import { ConstantString } from "../../../../common/constants";
+import { getResourceGroupNameFromResourceId } from "../../../../common/tools";
+import { ResourceGroupNotExistError } from "../../../../error/azure";
+import { DeployArmError, GetArmDeploymentError } from "../../../../error/arm";
+import { innerGetDeploymentError, innerGetDeploymentOperations } from "./innerHandleError";
+
+// constant string
+const ErrorCodes: { [key: string]: string } = {
+  InvalidTemplate: SolutionError.FailedToValidateArmTemplates,
+  InvalidTemplateDeployment: SolutionError.FailedToValidateArmTemplates,
+  ResourceGroupNotFound: SolutionError.ResourceGroupNotFound,
+};
+
+export type DeployContext = {
+  ctx: SolutionContext;
+  finished: boolean;
+  client: ResourceManagementClient;
+  resourceGroupName: string;
+  deploymentStartTime: number;
+  deploymentName: string;
+};
+
+/* eslint-disable @typescript-eslint/no-namespace */
+export namespace ArmErrorHandle {
+  export function fetchInnerError(error: any): any {
+    if (!error.details) {
+      return error;
+    }
+    if (error.details.error) {
+      return fetchInnerError(error.details.error);
+    } else if (error.details instanceof Array && error.details[0]) {
+      return fetchInnerError(error.details[0]);
+    }
+    return error;
+  }
+
+  export async function handleArmDeploymentError(
+    error: any,
+    deployCtx: DeployContext
+  ): Promise<Result<undefined, FxError>> {
+    // return the error if the template is invalid
+    if (Object.keys(ErrorCodes).includes(error.code)) {
+      if (error.code === "InvalidTemplateDeployment") {
+        error = fetchInnerError(error);
+      }
+      if (error.code === "ResourceGroupNotFound") {
+        return err(
+          new ResourceGroupNotExistError(
+            deployCtx.resourceGroupName,
+            deployCtx.client.subscriptionId
+          )
+        );
+      } else if (error.code === "InvalidTemplate" || error.code === "InvalidTemplateDeployment") {
```

Comment:
this "invalidtemplate" "invalidatetemplatedeployment" or "resourcegroupNotfound" may abstract contant strings. you may submit another PR later.